internal/wal/wal_aof.go

```diff
@@ -65,31 +170,212 @@ func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
-		slog.Warn("failed to write entry length to WAL", slog.Any("error", err.Error()))
+	buf := bufferPool.Get().([]byte)
+	buf = buf[:0]
+	defer bufferPool.Put(buf)
+
+	length := uint32(len(data))
+	totalSize := 4 + len(data)
+
+	if cap(buf) < totalSize {
+		buf = make([]byte, totalSize)
+	} else {
+		buf = buf[:totalSize]
+	}
+
+	binary.BigEndian.PutUint32(buf[0:4], length)
+	copy(buf[4:], data)
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	if w.rotationMode == "size" {
+		if err := w.rotateLogIfNeeded(); err != nil {
+			slog.Warn("failed to rotate WAL log", slog.Any("error", err.Error()))
+			return
+		}
 	}
-	writeBuf.Write(data)
 
-	if _, err := w.file.Write(writeBuf.Bytes()); err != nil {
+	if _, err := w.writer.Write(buf); err != nil {
 		slog.Warn("failed to write serialized command to WAL", slog.Any("error", err.Error()))
+		return
 	}
 
-	if err := w.file.Sync(); err != nil {
-		slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+	if w.writeMode == "fsync" {
+		if w.isBuffered && w.bufWriter != nil {
+			if err := w.bufWriter.Flush(); err != nil {
+				slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+				return
+			}
+		}
+		if err := w.file.Sync(); err != nil {
+			slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+			return
+		}
+	} else {
+		w.syncMutex.Lock()
+		defer w.syncMutex.Unlock()
+		if time.Since(w.lastSyncTime) >= w.syncInterval {
+			if w.isBuffered && w.bufWriter != nil {
+				if err := w.bufWriter.Flush(); err != nil {
+					slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+					return
+				}
+			}
+			if err := w.file.Sync(); err != nil {
+				slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+				return
+			}
+			w.lastSyncTime = time.Now()
+		}
 	}
 
 	slog.Debug("logged command in WAL", slog.Any("command", c.Repr()))
 }
 
-func (w *WALAOF) Close() error {
-	if w.file == nil {
+func (w *WALAOF) rotateLogIfNeeded() error {
+	fileInfo, err := w.file.Stat()
+	if err != nil {
+		return err
+	}
+
+	var bufferedSize int64 = 0
+	if w.isBuffered && w.bufWriter != nil {
+		bufferedSize = int64(w.bufWriter.Buffered())
+	}
+
+	if fileInfo.Size()+bufferedSize >= w.maxFileSize {
+		return w.rotateLog()
+	}
+	return nil
+}
+
+func (w *WALAOF) rotateLog() error {
+	if w.isBuffered && w.bufWriter != nil {
+		if err := w.bufWriter.Flush(); err != nil {
+			return err
+		}
+	}
+	if err := w.file.Sync(); err != nil {
+		return err
+	}
+	if err := w.file.Close(); err != nil {
+		return err
+	}
+
+	w.currentSegmentIndex++
+	if w.currentSegmentIndex >= w.maxSegments {
+		if err := w.deleteOldestSegment(); err != nil {
+			return err
+		}
+	}
+
+	segmentFileName := fmt.Sprintf("%s%d", segmentPrefix, w.currentSegmentIndex)
+	path := filepath.Join(w.logDir, segmentFileName)
+
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm not sure if O_TRUNC is the appropriate flag to use here. It reflects a possibility of an existing file being modified which is unacceptable for a WAL. In fact we should replace it with O_EXCL. It will ensure the operation will fail if the file already exists. If the file already exists it will return an error which we should handle here.

```suggestion
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0644)
if err != nil {
    if os.IsExist(err) {
        // Handle the case where the file already exists
        return fmt.Errorf("WAL segment already exists: %s", path)
    }
    return err
}
```